go-client/client/client.go

```diff
@@ -0,0 +1,262 @@
+		return ErrClosedClient
+	}
+
+	err := client.listFields(ctx, func(update *applicationpb2.FieldsChangeUpdate) {
+		client.handleFieldChanges(update)
+	})
+	if err != nil {
+		return err
+	}
+
+	if opt == FetchOnce {
+		client.appStub.cancelFunc()
+	}
+
+	return nil
+}
+
+// Returns a list of the (global) tables that can be opened with OpenTable.
+// This can be updated using FetchTables.
+func (client *Client) ListOpenableTables() []string {
+	client.tablesLock.Lock()
+	defer client.tablesLock.Unlock()
+
+	var result []string
+	for id := range client.tables {
+		if id.appId == "scope" {
+			result = append(result, id.fieldName)
+		}
+	}
+	return result
+}
+
+func (client *Client) newTicketNum() int32 {
+	client.nextTicket += 1
+	if client.nextTicket <= 0 {
+		// TODO:
+		panic("out of tickets")
+	}
+
+	return client.nextTicket
+}
+
+func (client *Client) newTicket() ticketpb2.Ticket {
+	id := client.newTicketNum()
+
+	return client.makeTicket(id)
+}
+
+func (client *Client) makeTicket(id int32) ticketpb2.Ticket {
+	bytes := []byte{'e', byte(id), byte(id >> 8), byte(id >> 16), byte(id >> 24)}
+
+	return ticketpb2.Ticket{Ticket: bytes}
+}
+
+// Sends an entire query over to the server all at once and returns the resulting tables.
+//
+// If this function completes successfully, the number of tables returned will always match the
+// number of query nodes passed.
+//
+// This may return a QueryError if the query is invalid.
+func (client *Client) ExecQuery(ctx context.Context, nodes ...QueryNode) ([]*TableHandle, error) {
+	if client.Closed() {
+		return nil, ErrClosedClient
+	}
+
+	return execQuery(client, ctx, nodes)
+}
+
+// Closes the connection to the server and frees any associated resources.
+// Once this method is called, the client and any TableHandles from it cannot be used.
+func (client *Client) Close() {
+	client.sessionStub.Close()
+	client.appStub.Close()
+	if client.grpcChannel != nil {
+		client.grpcChannel.Close()
+		client.grpcChannel = nil
+	}
+}
+
+// This is thread-safe
+func (client *Client) withToken(ctx context.Context) context.Context {
+	return metadata.NewOutgoingContext(context.Background(), metadata.Pairs("deephaven_session_id", string(client.getToken())))
+}
+
+// Directly uploads and executes a script on the deephaven server.
+// The script language depends on the scriptLanguage argument passed when creating the client.
+func (client *Client) RunScript(context context.Context, script string) error {
+	// This has to shadow the consoleStub method in order to handle the listfields loop
+
+	if client.Closed() {
+		return ErrClosedClient
+	}
+
+	restartLoop := client.appStub.isListing()
+	client.appStub.cancelListLoop()
+
+	if restartLoop {
+		// Clear out the table list to avoid any duplicate entries
+		client.tables = make(map[fieldId]*TableHandle)
+	}
+
+	err := client.consoleStub.RunScript(context, script)
+	if err != nil {
+		client.FetchTables(context, FetchRepeating) // At least try to restart this
```

Comment:
I'm unclear on what is happening and what should be.  This looks like:
1) script fails
2) we assume something may have happened during the failure, so we fetch the tables repeatedly
I assume this is to get new tables that may have been created via the incomplete script.

What if we are already fetching tables repeatedly?  Should this call happen?
What if the user never requested repeated updates?  Do they now get them no matter what?
What are the implications of these decisions?

Reply:
Fixed, there was a missing if-statement.
This is meant to continue the FetchTables call once RunScript has finished. Due to an implementation detail, FetchTables has to be canceled during RunScript, so it must be manually restored afterwards.